runtime/pkg/common/retryutil.go

```diff
@@ -0,0 +1,53 @@
+package common
+
+import (
+	"context"
+	"errors"
+	"math"
+	"math/rand"
+	"time"
+)
+
+const (
+	MaxSleepMillis  = 3000
+	BaseSleepMillis = 300
+)
+
+// Task is a function that returns an error if it fails
+type Task func() error
+
+// ShouldRetryFunc is a function that returns true if the error should be retried
+type ShouldRetryFunc func(error) bool
+
+// Retry executes the task with exponential backoff
+func Retry(ctx context.Context, task Task, shouldRetry ShouldRetryFunc, maxTries int) error {
+	if maxTries <= 0 {
+		return errors.New("maxTries must be greater than 0")
+	}
+	var err error
+	for try := 0; try < maxTries; try++ {
+		if ctx.Err() != nil {
+			return errors.Join(ctx.Err(), err)
+		}
+
+		err = task()
+		if err == nil {
+			return nil
+		}
+
+		if !shouldRetry(err) {
+			return err
+		}
+
+		sleepMillis := nextRetrySleepMillis(try)
+
+		time.Sleep(time.Duration(sleepMillis) * time.Millisecond)
```

Comment:
This has an issue that `ctx` cancellation will not cancel the sleep, so it may hang until retry before checking the context. Other places, we use the `github.com/eapache/go-resiliency/retrier` package for retries, might be worth using that instead – it has a `RunCtx` function which checks for cancellation during sleeping periods.

Reply:
changed to use that. only thing is the fuzzy multiplier here tries to prevents retry hotspots, not sure how practical it is though.